src/troika/sites/sge.py

```diff
@@ -0,0 +1,266 @@
+    (b'-o', b'foo')
+    >>> _split_sge_directive(b"-N job")
+    (b'-N', b'job')
+    >>> _split_sge_directive(b"-V")
+    (b'-V', None)
+    """
+    m = re.match(rb"(\S+)(\s+)?(.*)?$", arg)
+    if m is None:
+        raise RunError(r"Malformed qsub argument: {arg!r}")
+    key, sep, val = m.groups()
+    if sep is None:
+        assert val == b""
+        val = None
+    return key, val
+
+
+class SGEDirectiveParser(BaseParser):
+    """Parser that processes a script to extract SGE directives
+
+    Parameters
+    ----------
+    drop_keys: Iterable[bytes]
+        Directives to ignore, e.g. ``[b'-o', b'-e']``
+
+    Members
+    -------
+    data: collections.OrderedDict[bytes, (bytes or None, bytes)]
+        Directives that have been parsed. The first item of the dict value is
+        the parsed value, if any, and the second is the full line, including
+        the line terminator.
+    """
+
+    DIRECTIVE_RE = re.compile(rb"^#\s*$\s+(.+)$")
+
+    def __init__(self, drop_keys=None):
+        super().__init__()
+        self.data = OrderedDict()
+        if drop_keys is None:
+            drop_keys = []
+        self.drop_keys = set(drop_keys)
+
+    def feed(self, line):
+        """Process the given line
+
+        See ``BaseParser.feed``
+        """
+        m = self.DIRECTIVE_RE.match(line)
+        if m is None:
+            return False
+
+        key, value = _split_sge_directive(m.group(1))
+        if key not in self.drop_keys:
+            self.data[key] = (value, line)
+
+        return True
+
+
+def _translate_export_vars(value):
+    if value == b"all":
+        return b"-V"
+    if value == b"none":
+        return
+    return b"-v %s" % value
+
+
+def _translate_mail_type(value):
+    trans = {b"none": b"n", b"begin": b"b", b"end": b"e", b"fail": b"a"}
+    vals = value.split(b",")
+    newvals = []
+    for val in vals:
+        newval = trans.get(val.lower())
+        if newval is None:
+            _logger.warn("Unknown mail_type value %r", val)
+            newval = val
+        newvals.append(newval)
+    return b"-m %s" % b"".join(newvals)
+
+
+class SGESite(Site):
+    """Site managed using SGE"""
+
+
+    directive_prefix = b"#$ "
+    directive_translate = {
+        "billing_account": b"-A %s",
+        "error_file": b"-e %s",
+        "export_vars": _translate_export_vars,
+        "mail_type": _translate_mail_type,
+        "mail_user": b"-M %s",
+        "name": b"-N %s",
+        "output_file": b"-o %s",
+        "priority": b"-p %s",
+        "queue": b"-q %s",
+        "walltime": b"-l walltime=%s",
+    }
+
+    SUBMIT_RE = re.compile("(Your job )?(\d+)")
+
+    def __init__(self, config, connection, global_config):
+        super().__init__(config, connection, global_config)
+        self._qsub = command_as_list(config.get('qsub_command', 'qsub'))
+        self._qdel = command_as_list(config.get('qdel_command', 'qdel'))
+        self._qsig = command_as_list(config.get('qsig_command', 'qsig'))
```

Comment:
You are right. That was my conclusion as well, but with qsub -notify and trapping on SIGUSR1 you can at least trap the killing. I added -notify this in my site config, but could of course be a part of this Site class.

Reply:
Do we want to add this by default? Another option would be to add a directive, e.g. `"notify_end": b"-notify"` that the user may set if they want to get a signal. I don't mind too much since it is pretty harmless.

Reply:
I removed -notify to the qsub command for sge. However, I don't get it to work directly as a directive unless I specify the directive as:
`"notify_end": b"-notify %s",`
and in the job specify:
`#TROIKA notify_end=`

This is because the parsing expects key,value pairs and I have to handle an empty value. I could maybe add an extra translator or a pre-submit hook or something similar. What do you think?